sdk/core/azure-core/test/ut/base64.cpp

```diff
@@ -61,3 +62,52 @@ TEST(Base64, Basic)
       [](std::string expect, std::string actual) { return expect == actual; }, result, "AQIDBAUG");
   EXPECT_TRUE(std::equal(subsection.begin(), subsection.end(), Base64Decode(result).begin()));
 }
+
+static thread_local std::mt19937_64 random_generator(std::random_device{}());
+
+static char RandomChar()
+{
+  const char charset[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  std::uniform_int_distribution<std::size_t> distribution(0, sizeof(charset) - 2);
+  return charset[distribution(random_generator)];
+}
+
+void RandomBuffer(char* buffer, std::size_t length)
+{
+  char* start_addr = buffer;
+  char* end_addr = buffer + length;
+
+  const std::size_t rand_int_size = sizeof(uint64_t);
+
+  while (uintptr_t(start_addr) % rand_int_size != 0 && start_addr < end_addr)
+  {
+    *(start_addr++) = RandomChar();
+  }
+
+  std::uniform_int_distribution<uint64_t> distribution(0ULL, std::numeric_limits<uint64_t>::max());
+  while (start_addr + rand_int_size <= end_addr)
+  {
+    *reinterpret_cast<uint64_t*>(start_addr) = distribution(random_generator);
+    start_addr += rand_int_size;
+  }
+  while (start_addr < end_addr)
+  {
+    *(start_addr++) = RandomChar();
+  }
+}
+
+inline void RandomBuffer(uint8_t* buffer, std::size_t length)
+{
+  RandomBuffer(reinterpret_cast<char*>(buffer), length);
+}
+
+TEST(Base64, Roundtrip)
+{
+  for (std::size_t len : {0, 10, 100, 1000, 10000})
+  {
+    std::vector<uint8_t> data;
+    data.resize(len);
+    RandomBuffer(data.data(), data.size());
+    EXPECT_EQ(Base64Decode(Base64Encode(data)), data);
```

Comment:
out of curiosity, if this ever fail, how would we do reproduce it?  Since it is random generation, How do we know what was the data content?  
Maybe add a print to console to display the data raw data upon failure?  gtest would only display the Encoded and Decoded data right?

Reply:
Well, it's good but i don't feel it's necessary.

I don't expect the encoding/decoding impl will be changed frequently. Even it's changed and the test case fails, it's a clear message to the developer that the last change he's made is problematic. He could find some other way to debug.
So I think improving this piece of test code is some kind of investment that we can hardly benefit from.

Reply:
I am not a fan of using random/non-deterministic test data for this exact reason. The input length in some cases is 10K characters, so printing to a console might not be the best.

That said, gtest's `EXPECT_EQ` is showing us the data since that is what we are comparing the round-tripped transformed result against. We pass-in `data` as a parameter. Therefore, we don't need to print that to console ourselves:
```text
..\..\..\sdk\core\azure-core\test\ut\base64.cpp(112): error: Expected equality of these values:
      Base64Decode(Base64Encode(wrong))
        Which is: { '\x1' (1), '\x2' (2), '\x3' (3) }
      data
        Which is: { '\xC0' (192), '\x82' (130), '\x89' (137), '{' (123, 0x7B), '\xEE' (238), '\xA5' (165), '\xAB' (171), '\x99' (153), 'A' (65, 0x41), '6' (54, 0x36) }
```